source/log.c

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright 2010 - 2018 Amazon.com, Inc. or its affiliates.All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file.This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied.See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/condition_variable.h>
+#include <aws/common/linked_list.h>
+#include <aws/common/log.h>
+//#include <aws/common/memory_pool.h>
+#include <aws/common/mutex.h>
+#include <aws/common/thread.h>
+
+#include <inttypes.h>
+#include <stdio.h>
+#include <time.h>
+
+struct s_msg {
+    struct aws_linked_list_node node;
+    const char *tag;
+};
+
+struct s_log_ctx {
+    struct aws_allocator *alloc;
+    void *msg_pool;
+    size_t message_size;
+    struct aws_linked_list messages;
+};
+
+/* Logging globals. */
+struct s_log_ctx *s_ctx;
+struct aws_mutex s_mutex = AWS_MUTEX_INIT;
+aws_log_report_fn *s_report;
+enum aws_log_level s_level;
+
+/* Logging thread state. */
+bool s_log_thread_running;
+struct aws_thread s_log_thread;
+struct aws_condition_variable s_cv = AWS_CONDITION_VARIABLE_INIT;
+struct aws_mutex s_cv_mutex = AWS_MUTEX_INIT;
+
+static inline struct s_msg *s_msg_new(void) {
+    if (!s_ctx) {
+        return NULL;
+    }
+
+    aws_mutex_lock(&s_mutex);
+    struct s_msg *msg = (struct s_msg *)aws_mem_acquire(s_ctx->alloc, sizeof(struct s_msg) + s_ctx->message_size);
```

Comment:
we assume the thread safety of allocators. This is usually the case anyways since most allocators are stateless and they pin any state they need to the memory they allocate anyways. If they're more complicated than that, it's up to the allocator to handle it.